notebooks/Autoencoders.ipynb

```diff
@@ -0,0 +1,632 @@
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "id": "iT-ibpi7voxQ"
+   },
+   "source": [
+    "### Table of contents"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "id": "ekJWKPFMvoxR"
+   },
+   "source": [
+    "* [Data](#Data)\n",
+    "* [Define Model](#Define-Keras-Model)\n",
+    "* [Training](#Training)\n",
+    "* [Explore Results](#Explore-Results)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "id": "6avb3GBQDQyG"
+   },
+   "source": [
+    "Install SciKeras"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "id": "qCcyTjVkvoxR"
+   },
+   "outputs": [],
+   "source": [
+    "!python -m pip install git+https://github.com/adriangb/scikeras.git@master"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "id": "EZveNcetDQyL"
+   },
+   "source": [
+    "Silence TensorFlow warnings to keep output succint."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 22,
+   "metadata": {
+    "id": "ekNmO_GPDQyL"
+   },
+   "outputs": [],
+   "source": [
+    "import warnings\n",
+    "from tensorflow import get_logger\n",
+    "get_logger().setLevel('ERROR')"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "id": "hCuOBH8AvoxX"
+   },
+   "source": [
+    "## Data"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "id": "i3fAUKBUvoxY"
+   },
+   "source": [
+    "We load the dataset from the Keras tutorial. The dataset consists of images of cats and dogs."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 23,
+   "metadata": {
+    "colab": {
+     "base_uri": "https://localhost:8080/",
+     "height": 51
+    },
+    "id": "QM74xeoe-1S-",
+    "outputId": "23946eec-c32f-4db6-8bda-e12bda6492ea"
+   },
+   "outputs": [
+    {
+     "output_type": "stream",
+     "name": "stdout",
+     "text": [
+      "(60000, 784)\n(10000, 784)\n"
+     ]
+    }
+   ],
+   "source": [
+    "from tensorflow.keras.datasets import mnist\n",
+    "import numpy as np\n",
+    "\n",
+    "\n",
+    "(x_train, _), (x_test, _) = mnist.load_data()\n",
+    "x_train = x_train.astype('float32') / 255.\n",
+    "x_test = x_test.astype('float32') / 255.\n",
+    "x_train = x_train.reshape((len(x_train), np.prod(x_train.shape[1:])))\n",
+    "x_test = x_test.reshape((len(x_test), np.prod(x_test.shape[1:])))\n",
+    "print(x_train.shape)\n",
+    "print(x_test.shape)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "id": "yuR10hymK0dh"
+   },
+   "source": [
+    "## Define Keras Model"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "We will be defining a very simple autencoder. We define _three_ model building methods:\n",
```

Comment:
Is it necessary to define three functions? Could you do something like below?

``` python
def get_all_models(encoding_dim: int) -> Tuple[Model, Model, Model]
    input_img = keras.Input(shape=(784,), name="input")
    encoded = keras.layers.Dense(encoding_dim, activation='relu', name="encoded")(input_img)
    decoded = keras.layers.Dense(784, activation='sigmoid', name="output")(encoded)
    autoencoder_model = keras.Model(input_img, decoded)
    return autoencoder_model, encoded, decoded

def get_fit_model():
    return get_all_models()[0]

def get_model():
    return get_all_models()[1]

def get_inv_model():
    return get_all_models()[2]
```

Reply:
Unfortunately, they need to share the same layer/object. The simplest I can think of is:

```python3
encoding_dim = 32  # 32 floats -> compression of factor 24.5, assuming the input is 784 floats
input_img = keras.Input(shape=(784,))
encoded = keras.layers.Dense(encoding_dim, activation="relu")(input_img)
decoded = keras.layers.Dense(784, activation="sigmoid")(encoded)
autoencoder_model = keras.Model(input_img, decoded)
encoder_model = keras.Model(input_img, encoded)
encoded_input = keras.Input(shape=(encoding_dim,))
decoder_layer = autoencoder_model.layers[-1]
decoder_model = keras.Model(encoded_input, decoder_layer(encoded_input))

def get_fit_model():
    return autoencoder_model

def get_tf_model():
    return encoder_model

def get_inverse_tf_model():
    return decoder_model
```

But that looses the ability to set the encoding dimensions as a parameter.

Reply:
Why not have a flag to determine the return value?

``` python
def get_coder(encoding_dim: int, out: str = "autoencoder") -> Tuple[Model, Model, Model]
    input_img = keras.Input(shape=(784,), name="input")
    encoded = keras.layers.Dense(encoding_dim, activation='relu', name="encoded")(input_img)
    decoded = keras.layers.Dense(784, activation='sigmoid', name="output")(encoded)
    autoencoder = keras.Model(input_img, decoded)

    returns = {"autoencoder": autoencoder, "encoder": encoded, "decoder": decoded}
    return returns[out]

# ...

fit_model = get_all_models(model=get_coder, model__out="autoencoder")
compress = get_all_models(model=get_coder, model__out="encoder")
expand = get_all_models(model=get_coder, model__out="decoder")
```

Reply:
Every time `get_coder` is called, it would return a different `encoded` object. There would be nothing tying the autoencoder model (which is what is being trained) to the `encoded` and `decoded` layers used for inference. Unless I am missing something?

Reply:
We could do something like:

```python3
def get_fit_model(encoding_dim: int, out: str) -> keras.Model:
    input_img = keras.Input(shape=(784,))
    encoded = layers.Dense(encoding_dim, activation="relu")(input_img)
    decoded = layers.Dense(784, activation="sigmoid")(encoded)
    autoencoder_model = keras.Model(input_img, decoded)
    encoder_model = keras.Model(input_img, encoded)
    encoded_input = keras.Input(shape=(encoding_dim,))
    decoder_layer = autoencoder_model.layers[-1]
    decoder_model = keras.Model(encoded_input, decoder_layer(encoded_input))

    returns = {"autoencoder": autoencoder_model, "encoder": encoder_model, "decoder": decoder_model}
    out_new = None
    while True:
        out_new = yield returns[out if out_new is None else out_new]

gen = get_fit_model(32, "autoencoder")

a = next(gen)
b = gen.send("decoder")
c = next(gen)
``` 

But that would require making SciKeras aware of generators, `send`, etc.

Reply:
> Every time get_coder is called, it would return a different encoded and decoded object.

Only because Keras uses a different random state to create the encoder/decoder models, correct? Does Keras have scheme to initialize the models from the same random state?

Reply:
It's not really random state related. It's because the layers have to be connected, as in, connected in the tensorflow graph, so that when you train the auto-encoder the weights in the encoder/decoder are updated / the same. You could create a completely new model with the same architecture and just copy over the weights, but you would still need a reference to the autoencoder's trained layers (to get the weights) so it's no different really.

Reply:
In other words, if `==` was overloaded to compare architecture _and_ weights, what we need is:

`autoencoder_model[encode_layers] == encoder_model`

Which can most easily be achieved if we satisfy:

`autoencoder_model[encode_layers] is encoder_model`

Reply:
@stsievert let me know if I am still not being clear enough, I can make a small example or something

Reply:
Eh, I haven't commented because I don't think this is important enough to worry about.

Reply:
Got it, thank you. Do you plan on taking another look at the PR as a whole, or should I move it forward?